enterprise/backend/src/metabase_enterprise/caching/api.clj

```diff
@@ -0,0 +1,118 @@
+                                                         [:= :model_id :report_card.id]
+                                                         [:= :report_card.collection_id collection]]
+                                      :report_dashboard [:and
+                                                         [:= :model [:inline "dashboard"]]
+                                                         [:= :model_id :report_dashboard.id]
+                                                         [:= :report_dashboard.collection_id collection]]]
+                          :where     [:case
+                                      [:= :model [:inline "question"]]  [:!= :report_card.id nil]
+                                      [:= :model [:inline "dashboard"]] [:!= :report_dashboard.id nil]
+                                      :else                             true]})]
+    {:items (for [item items]
+              {:model    (:model item)
+               :model_id (:model_id item)
+               :strategy (assoc (:config item) :type (:strategy item))})}))
+
+(api/defendpoint PUT "/"
+  "Store cache configuration."
+  [:as {{:keys [model model_id strategy]} :body}]
+  {model    [:enum "root" "database" "dashboard" "question"]
+   model_id ms/IntGreaterThanOrEqualToZero
+   strategy [:and
+             [:map
+              [:type [:enum :nocache :ttl :duration :schedule :query]]]
+             [:multi {:dispatch :type}
+              [:nocache  [:map
+                          [:type keyword?]]]
+              [:ttl      [:map
+                          [:type keyword?]
+                          [:multiplier ms/PositiveInt]
+                          [:min_duration ms/PositiveInt]]]
+              [:duration [:map
+                          [:type keyword?]
+                          [:duration ms/PositiveInt]
+                          [:unit [:enum "hours" "minutes" "seconds" "days"]]]]
+              [:schedule [:map
+                          [:type keyword?]
+                          [:schedule u.cron/CronScheduleString]]]
+              [:query    [:map
+                          [:type keyword?]
+                          [:field_id int?]
+                          [:aggregation [:enum "max" "count"]]
+                          [:schedule u.cron/CronScheduleString]]]]]}
+  (validation/check-has-application-permission :setting)
+  (assert-valid-models model [model_id])
+  (t2/with-transaction [_tx]
+    (let [criteria {:model    model
+                    :model_id model_id}
+          data     {:strategy (:type strategy)
+                    :config   (dissoc strategy :type)}
+          current  (t2/select-one :model/CacheConfig :model model :model_id model_id {:for :update})]
+      {:id (u/prog1 (if current
+                      (first (t2/update-returning-pks! :model/CacheConfig criteria data))
+                      (t2/insert-returning-pk! :model/CacheConfig (merge criteria data)))
```

Comment:
why do we want the ability to create in PUT API here?

also, if it's creating a cache config, we shouldn't audit it.

Reply:
It's 100% idiomatic for PUT to upsert:

> The HTTP PUT request method creates a new resource or replaces a representation of the target resource with the request payload.
> 
> The difference between PUT and [POST](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST) is that PUT is idempotent: calling it once or several times successively has the same effect (that is no side effect), whereas successive identical [POST](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST) requests may have additional effects, akin to placing an order several times.[^1]

That said, the semantics of this handler are currently PATCH[^2], unless we add the cache keys to the route.

[^1]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PUT
[^2]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PATCH

Reply:
PATCH support has always been patchy, using PUT instead is widely adopted abuse.

I hadn't realized that the parameters could be vectors however... if we're mutating multiple resources I feel that this should actually be POST. I don't think there are any hard and fast rules about expected POST semantics, it's the catch all for RPC.